### ISSUE TITLE
Fix Task Library Async Causality Stack Reconstruction

### ISSUE THREAD

**Issue by vancem** [dotnet/runtime#11474]
In synchronous code, a stack trace is a super-powerful way of describing 'where' you are in a program in a very useful way.   We want this ability for asynchronous code as well.   However this is not trivial because threads (and therefore their stacks) are constantly being reused.    However the concept that one method 'calls' (more generally 'causes') another method to do something (its callee), is still valid 

We have events (Scheduled, Begin, End),  that fire when tasks are used for concurrency, and other events (WaitBegin and WaitEnd), that are used when C# 'await' routines are used.   We also have logic in the TraceEvent package that knows how to take these events (and the stacks that are on some of them) and form 'causality stack' from them. 

There are reason, however to believe that there are 'holes' in this mechanism.   That is there are ways of using the Task library that do not log sufficient events (or perhaps TraceEvent does not recognize them), and thus causality is broken.

This issue is meant to track the end-to-end scenario.  The end goal is that tools like AppInsights Profiler and PerfView (which both use the TraceEvent library), have sufficient information to always generate a useful 'stack trace' (causality trace).   

- [ ] Create tests that insure we can do this causality stack reconstruction, including
        - [] Basic ASP.NET Core scenarios (e.g. MVC) at various points
        - [] More methodical testing that tries to use all the different APIs of Task and features of C# async 
              and insure that we can get good stacks when waiting with these APIs being used to set up
              the causality chain.  
- [] We go to some trouble to minimize the number of events logged by the task library.  There is a good
      chance that this is what is causing some of the bugs.    It would be useful to have a low level 
      mechanism that is only turned on for testing that does not worry about efficiency, but logs at
      precisely the right time.    This makes it easy to do the stack 'both ways' and then compare that they 
      are the same.   Thus we can take 'arbitrary' code (say all the ASP.NET tests) and run it through 
      this harness that turns on these extra events and does the comparison of what we get 
      to the 'gold standard' events, so we can find new problems (because we did not have code 
      coverage, and add more targeted tests when we find new things.  
- [] Today our TaskWait event does not indicate whether a task is immediately waiting on another task.  
      This often happens when you have methods calling other async methods.    It is VERY useful to
      indicate in the GUI that the time that all these nested Tasks are waiting on is EXPECTED to overhap
      (and thus just count it once).   Today we use heuristics, we should fix it so that we simply say so 
      in the event (and thus can give a better display to the user).   
- [] Fix any issues that come up from that testing.  

Note that our current approach is relatively expensive in that it needs two events to be fired for every blocking call.    There may be ways to do better than this (by leveraging the information in the Task objects), so that is worth thinking about as part of this.
